### ISSUE TITLE
Memory leak when deleting root commands

### ISSUE THREAD

**Issue by Mr-EmPee** [Incendo/cloud#435]
When you delete a root command some references to it are kept.
I'm using cloud-paper and cloud-annotations 1.8.3 on a 1.19.3 server

![Heap Dump](https://user-images.githubusercontent.com/86979852/227771899-577e3afd-1b16-46a0-8e17-14c3147b48dc.PNG)

**Comment by rgnter** [comment]
That's minecraft keeping a reference in the brigadier dispatcher. You can't delete that. It's going to be deleted on reload or restart.

**Comment by Mr-EmPee** [comment]
@rgnter Why isn't that deletable? Do you mean through API?

**Comment by rgnter** [comment]
When plugins are enabled, _bukkit_ commands are merged with _vanilla_ commands, and stored away in command _dispatcher_. This dispatched is not available through the API.

**Comment by Mr-EmPee** [comment]
My question is why can't we just access the map with reflection and remove the command?

**Comment by rgnter** [comment]
Because using reflection is inefficient, burdensome(from maintenance point of view) and slow. Let them keep the reference, it's not that much trouble for us. 
I can't see any real reason why to remove nodes. If there's something I can't see, enlighten me.

**Comment by Mr-EmPee** [comment]
Okay that's understandable but we should at least break the chain that links the command class (Where you define the command annotations) to the BukkitCommand, bc if the chain hasn't been broken that class can't be garbage collected and that could cause a big memory leak depending on how someone has structured their code.

**Comment by rgnter** [comment]
I like how you think, but we can't. It's impossible
- BukkitCommand is the only way to register commands in the API. 
- API doesn't provide a direct layer to the Brigadier API. 
  - All commands registered to Brigadier API are treated as vanilla commands (appending minecraft namespace in front of them)  
  - We could to it hacky-way, but that's just irresponsible and could cause undefined behavior

**Comment by Mr-EmPee** [comment]
What I meant by breaking the chain was making this field https://github.com/Incendo/cloud/blob/a4cc749b91564af57bb7bba36dd8011b556c2b3a/cloud-annotations/src/main/java/cloud/commandframework/annotations/MethodCommandExecutionHandler.java#L52
null

**Comment by rgnter** [comment]
So, we can't make it null, because the node doesn't know when it's removed. And if it knew it's removed, how can we be sure that the node is not going to be added elsewhere.

**Comment by Mr-EmPee** [comment]
> And honestly, the "leak" is trivial. Nobody really removes nodes. And if, the memory footprint wouldn't be that huge to call it a "leak" (couple of hundreds bytes in HEAP). It is irresponsible, but we can't do much about it, unless we get direct access to the brigadier API.

The leak could be big depending on how your code is structured. For example, you just need to have a cache that stores some data or a reference to an instance that holds and manages a cache. 

I think that this problem shouldn't be overlooked or at least should be mentioned in the doc of the deleteRootCommand bc
I would expect (and probably other dev too) that if there are no references in my code to an instance and after doing a deleteRootCommand to be garbage collected.

> So, we can't make it null, because the node doesn't know when it's removed. And if it knew it's removed, how can we be sure that the node is not going to be added elsewhere.

That's true this means that we should provide a way to "invalidate" the MethodCommandExecutionHandler.
I want to state that idk internally how cloud-annotations or cloud-core works and I may be missing some important concepts.

**Comment by rgnter** [comment]
I see your concern, but hacking our way through the API just to fix one small issue for Bukkit is kinda-sorta bad. The issue only occurs when deleting a node which is based on annotations (API for node removal is also experimental). Lesson learned: decouple your logic and data. 

We probably can't do much other than document it.

**Comment by Mr-EmPee** [comment]
I don't think that hacking is the only way to fix this. There are other ways as I stated above

**Comment by rgnter** [comment]
> I don't think that hacking is the only way to fix this. There are other ways as I stated above

By hacking I meant introducing new experimental API to release the method handle. 😀 **There's really no way for us to tell when to safely release the method handle.**

You're right that this situation is bad and irresponsible of us to ignore, but when user of the library - in this case Bukkit, is holding a reference we can't so much about it. Bukkit never intended for commands to be removed. 

Imagine case where you provide a annotation based command, and another internal mechanism takes the reference to that command (not just bukkit) and you decide to release the method handle just because you removed the root node. 
Okay, Bukkit is not going to use it, but what about that other internal mechanism? We could cause something worse. 

**solution**: Disable the removal API for nodes on the Bukkit platform. Until direct interoperability with Brigadier is possible.

**Comment by rgnter** [comment]
Although I might have a simple solution to this. Once I'm available I'll try to craft something.

**Comment by Mr-EmPee** [comment]
I was thinking to wrap the MethodHandle inside a WeakReference, this was everyone is forced to make the check

**Comment by Mr-EmPee** [comment]
Btw, I've just seen that the CommandMethodContext has references to the instance as well

**Comment by OOP-778** [comment]
Hello, has this been fixed since I just got to this.

My proposed fix is in the BukkitRegistrationHandler, it uses the player#syncCommands to sync commands to players, but that doesn't update dispatcher, to update the dispatcher you can use CraftServer#syncCommands() method that will update the dispatcher and players. Need some reflection, but that's how I got it to reload properly.

Also noticed that not all aliases are being removed, only the namespaced ones from the knownCommandMap
so if I have a command bp under a plugin neo, it removes neo:bp, but bp itself does not get removed if it's an aliase.
